java/ql/src/semmle/code/java/frameworks/apache/Collections.qll

```diff
@@ -138,127 +188,420 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+      ["org.apache.commons.collections4", "org.apache.commons.collections"] +
+        [
+          // Note that when lambdas are supported we should have a model for collect, forAllButLastDo, forAllDo, transform
+          ";CollectionUtils;true;addAll;(Collection,Object[]);;ArrayElement of Argument[1];Element of Argument[0];value",
+          ";CollectionUtils;true;addAll;(Collection,Enumeration);;Element of Argument[1];Element of Argument[0];value",
+          ";CollectionUtils;true;addAll;(Collection,Iterable);;Element of Argument[1];Element of Argument[0];value",
+          ";CollectionUtils;true;addAll;(Collection,Iterator);;Element of Argument[1];Element of Argument[0];value",
+          ";CollectionUtils;true;addIgnoreNull;;;Argument[1];Element of Argument[0];value",
+          ";CollectionUtils;true;collate;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;collate;;;Element of Argument[1];Element of ReturnValue;value",
+          ";CollectionUtils;true;disjunction;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;disjunction;;;Element of Argument[1];Element of ReturnValue;value",
+          ";CollectionUtils;true;emptyIfNull;;;Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;extractSingleton;;;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;find;;;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Iterator,int);;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Iterable,int);;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Map,int);;MapKey of Argument[0];MapKey of ReturnValue;value",
+          ";CollectionUtils;true;get;(Map,int);;MapValue of Argument[0];MapValue of ReturnValue;value",
+          ";CollectionUtils;true;get;(Object,int);;ArrayElement of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Object,int);;Element of Argument[0];ReturnValue;value",
+          ";CollectionUtils;true;get;(Object,int);;MapKey of Argument[0];MapKey of ReturnValue;value",
+          ";CollectionUtils;true;get;(Object,int);;MapValue of Argument[0];MapValue of ReturnValue;value",
+          ";CollectionUtils;true;getCardinalityMap;;;Element of Argument[0];MapKey of ReturnValue;value",
+          ";CollectionUtils;true;intersection;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;intersection;;;Element of Argument[1];Element of ReturnValue;value",
+          ";CollectionUtils;true;permutations;;;Element of Argument[0];Element of Element of ReturnValue;value",
+          ";CollectionUtils;true;predicatedCollection;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;removeAll;;;Element of Argument[0];Element of ReturnValue;value",
+          ";CollectionUtils;true;retainAll;;;Element of Argument[0];Element of ReturnValue;value",
```

Comment:
I think we should model things like this pulling taint from `Argument[-1]` not `Argument[0]`, mainly because it's more explicable to the user seeing something tainted in the container itself and supposing it might get removed / retained than the reverse. Of course really we want an intersection, but our modelling isn't detailed enough for that.

Reply:
Since these are static methods, isn't `Arugment[0]` the tainted container?

Reply:
Doh right, yes. Ignore these.